conformance/utils/http/http.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"net/url"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+	"sigs.k8s.io/gateway-api/conformance/utils/roundtripper"
+)
+
+// ExpectedResponse defines the response expected for a given request.
+type ExpectedResponse struct {
+	Request    ExpectedRequest
+	StatusCode int
+	Backend    string
+	Namespace  string
+}
+
+// ExpectedRequest can be used as both the request to make and a means to verify
+// that echoserver received the expected request.
+type ExpectedRequest struct {
+	Host    string
+	Method  string
+	Path    string
+	Headers map[string]string
+}
+
+// MakeRequestAndExpectResponse makes a request with the given parameters and
+// verifies the response matches the provided ExpectedResponse.
+func MakeRequestAndExpectResponse(t *testing.T, r roundtripper.RoundTripper, gwAddr string, expected ExpectedResponse) {
```

Comment:
Have you considered having retry here in case the resources are being stabilized while this runs?
Or should these be run only once the status of resources have been verified (and thus ensures a deterministic state).

Reply:
These tests wait for the appropriate ready status to be set on the Route and Gateway. I think that should be interpreted as the Gateway being ready to handle these requests. Although it's possible that some implementations are setting status prior to proxies being completely updated, I think we should try to keep conformance tests as tight as possible initially so that users can trust a ready status across all conformant implementations.

Reply:
Okay. That's fine. We can get feedback and tweak this if needed.